cpp/src/io/fst/agent_dfa.cuh

```diff
@@ -0,0 +1,675 @@
+  }
+
+  template <int32_t BYPASS_LOAD,
+            typename SymbolMatcherT,
+            typename TransitionTableT,
+            typename CallbackOpT>
+  __device__ __forceinline__ void GetThreadStateTransitions(
+    SymbolMatcherT const& symbol_matcher,
+    TransitionTableT const& transition_table,
+    CharT const* d_chars,
+    OffsetT const block_offset,
+    OffsetT const num_total_symbols,
+    StateIndexT& state,
+    CallbackOpT& callback_op,
+    cub::Int2Type<BYPASS_LOAD> /**/)
+  {
+    using StateTransitionOpT = StateTransitionOp<CallbackOpT, TransitionTableT>;
+
+    // Start parsing and to transition states
+    StateTransitionOpT transition_op(transition_table, state, callback_op);
+
+    // Load characters into shared memory
+    if (!BYPASS_LOAD) LoadBlock(d_chars, block_offset, num_total_symbols);
+
+    // If this is a full block (i.e., all threads can parse all their symbols)
+    OffsetT num_block_chars = num_total_symbols - block_offset;
+    bool is_full_block      = (num_block_chars >= SYMBOLS_PER_BLOCK);
+
+    // Ensure characters have been loaded
+    __syncthreads();
+
+    // Thread's symbols
+    CharT* t_chars = &temp_storage.chars[threadIdx.x * SYMBOLS_PER_THREAD];
+
+    // Initialize callback
+    callback_op.Init(block_offset + threadIdx.x * SYMBOLS_PER_THREAD);
+
+    // Parse thread's symbols and transition the state-vector
+    if (is_full_block) {
+      GetThreadStateTransitions<SYMBOLS_PER_THREAD>(
+        symbol_matcher, t_chars, num_block_chars, transition_op, cub::Int2Type<true>());
+    } else {
+      GetThreadStateTransitions<SYMBOLS_PER_THREAD>(
+        symbol_matcher, t_chars, num_block_chars, transition_op, cub::Int2Type<false>());
+    }
+
+    callback_op.TearDown();
+  }
+};
+
+template <bool IS_TRANS_VECTOR_PASS,
+          bool IS_SINGLE_PASS,
+          typename DfaT,
+          typename TileStateT,
+          typename AgentDFAPolicy,
+          typename SymbolItT,
+          typename OffsetT,
+          typename StateVectorT,
+          typename OutOffsetScanTileState,
+          typename TransducedOutItT,
+          typename TransducedIndexOutItT,
+          typename TransducedCountOutItT>
+__launch_bounds__(int32_t(AgentDFAPolicy::BLOCK_THREADS)) __global__
+  void SimulateDFAKernel(DfaT dfa,
+                         SymbolItT d_chars,
+                         OffsetT const num_chars,
+                         StateIndexT seed_state,
+                         StateVectorT* __restrict__ d_thread_state_transition,
+                         TileStateT tile_state,
+                         OutOffsetScanTileState offset_tile_state,
+                         TransducedOutItT transduced_out_it,
+                         TransducedIndexOutItT transduced_out_idx_it,
+                         TransducedCountOutItT d_num_transduced_out_it)
+{
+  using AgentDfaSimT = AgentDFA<AgentDFAPolicy, SymbolItT, OffsetT>;
+
+  static constexpr int32_t NUM_STATES = DfaT::MAX_NUM_STATES;
+
+  enum {
```

Comment:
Is it worth to specify the underlying type here?

Reply:
I was thinking of something like `enum: uint32_t` or `enum class : uint32_t` but the constexpr option is also fine, probably even better.

Reply:
Yup, thought so. Thanks for clarifying, though. I just found `constexpr` more appropriate here.